### ISSUE TITLE
ZIO Test: Cancel after first failure feature

### ISSUE THREAD

**Issue by reibitto** [zio/zio#3280]
I particularly find this feature useful when I just refactor some code that potentially could break a lot of stuff. I do a test run as a sanity check. And if I see a failure, I'd rather stop there and fix the issue rather than have the tests continue to run for a while only to see something like "300 errors".

Technically this could be generalized as "Cancel after N failures". I don't know if that's useful or not, but could be in certain cases like "I want to fix the first 5 failing tests and then concentrate on the others later" or something. I don't know... just a thought.

Also, I imagined this being a TestArg, though I guess it could be a TestAspect too.

**Comment by reibitto** [comment]
One more thing to think about is failures when running tests in parallel. For example, scalatest has `CancelAfterFailure` but you can't use that with `ParallelTestExecution`.

I don't foresee ZIO Test having this problem considering ZIO supports interruption, but thought I'd mention it anyway.

**Comment by hearnadam** [comment]
I don't think this will be pursued. For me, this would reduce the usability of zio-test significantly. Maybe it's possible with an annotation?

**Comment by reibitto** [comment]
I didn't mean for this to be the default behavior. Just that it could be flipped on if someone wanted it. Sort of like how scalatest has the `CancelAfterFailure` trait. The only difference being that a cake-y solution like that isn't appropriate for zio-test, so I was imagining a TestAspect or runner flag like `test -- --cancel-after-failure` (and a shortcut flag) or something like that.

But if there's not much demand for this I'm fine with it saying closed. It's a fairly niche thing.
